### PR TITLE
keep stacktrace in abstract_conv optimizations

### LINKED ISSUE #3589: Keep creation stack trace during opt: abstract conv 2d CPU
Sub-ticket of #3018.
Use `copy_stack_trace` for optimizations defined in `theano/tensor/nnet/opt.py` (previously in `theano/tensor/nnet/abstract_conv2d.py`), and add checks that the stack trace is still there during tests.
Tests are now in `theano/sandbox/gpuarray/tests/test_abstractconv.py`.

### PR DESCRIPTION
**Pull request by vmichals**
fix #3589

### CODE REVIEW

File: theano/tensor/nnet/opt.py [hunk 1]
```diff
@@ -92,6 +94,7 @@ def local_abstractconv_gradweight_gemm(node):
     if node.op.filter_flip:
```
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: I think we should also add the stack trace on the output of `CorrMM_gradWeights`, the `rval` defined on l. 91.

File: theano/tensor/nnet/opt.py [hunk 2]
```diff
@@ -175,6 +181,7 @@ def local_conv2d_gradweight_cpu(node):
         rval = rval[:, :, ::-1, ::-1]
         rval = theano.tensor.patternbroadcast(rval,
                                               node.outputs[0].broadcastable)
+        copy_stack_trace(node.outputs[0], rval)
```
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: I would keep this one, but also add the stack trace to the output of `convGrad3D`, defined on l.172.

File: theano/tensor/nnet/opt.py [hunk 3]
```diff
@@ -245,6 +252,8 @@ def local_conv2d_gradweight_cpu(node):
         res = res[:, :, ::-1, ::-1]
 
     res = theano.tensor.patternbroadcast(res, node.outputs[0].broadcastable)
+
+    copy_stack_trace(node.outputs[0], res)
```
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: Again, this one is fine, but I would also add the trace to the `res` defined on line 249.

File: theano/tensor/nnet/opt.py [hunk 4]
```diff
@@ -278,6 +287,8 @@ def local_conv2d_gradinputs_cpu(node):
         rval = rval.dimshuffle(0, 4, 1, 2)
         rval = theano.tensor.patternbroadcast(rval,
                                               node.outputs[0].broadcastable)
+
+        copy_stack_trace(node.outputs[0], rval)
```
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: That one is fine as well, just add it to the output of `convTransp3D` as well.

File: theano/tensor/nnet/opt.py [hunk 5]
```diff
@@ -327,6 +338,7 @@ def local_conv2d_gradinputs_cpu(node):
                  direction_hint='bprop inputs')
     din = din(topgrad, filters)
```
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: Let's add the stack trace here as well.

File: theano/sandbox/gpuarray/tests/test_abstractconv.py [hunk 6]
```diff
@@ -75,6 +75,7 @@ def run_fwd(self, inputs_shape, filters_shape, ref=dnn_conv,
                         filter_flip=filter_flip,
                         input_shape=imshp,
                         filter_shape=kshp)
+        self.assertTrue(hasattr(c.tag, 'trace'))
         f_ref = theano.function([], c_ref, mode=mode)
         f = theano.function([], c, mode)
```
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_abstractconv.py]: Actually, you should check the presence of the stack trace on the output of the optimized graph, not the original variable.
The variable corresponding to the optimized output would be `f.maker.fgraph.outputs[0]`, and we want `f.maker.fgraph.outputs[0].tag.trace` to be there.

File: theano/sandbox/gpuarray/tests/test_abstractconv.py [hunk 7]
```diff
@@ -113,6 +114,7 @@ def run_gradweight(self, inputs_shape, filters_shape, output_shape,
                                             filter_flip=filter_flip,
                                             subsample=subsample,
                                             imshp=imshp, kshp=kshp)
+        self.assertTrue(hasattr(c.tag, 'trace'))
         c = c(inputs, output, filters_shape[-2:])
         c_ref = ref(inputs, output,
                     filters_shape,
```
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_abstractconv.py]: Same here (after defining `f` on line 124).

File: theano/sandbox/gpuarray/tests/test_abstractconv.py [hunk 8]
```diff
@@ -159,6 +161,7 @@ def run_gradinput(self, inputs_shape, filters_shape, output_shape, ref=dnn_gradi
                                            subsample=subsample,
                                            filter_flip=filter_flip,
                                            imshp=imshp, kshp=kshp)
+        self.assertTrue(hasattr(c.tag, 'trace'))
```
Comment (lamblin) [inline theano/sandbox/gpuarray/tests/test_abstractconv.py]: Same here (`f` on line 169).

**Comment by lamblin** [comment]
Hmm, you seemed to have had some trouble rebasing. We can try to solve that tomorrow if you want.

**Comment by vmichals** [comment]
I think I fixed it by hard resetting to the previous commit. Then I rebased and resolved a minor merge conflict (keeping changes from another commit, that was made after branching).

File: theano/tensor/nnet/opt.py (hunk 1 above)
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: (done)

File: theano/tensor/nnet/opt.py (hunk 2 above)
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: (done)

File: theano/tensor/nnet/opt.py (hunk 3 above)
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: (done)

File: theano/tensor/nnet/opt.py (hunk 4 above)
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: (done)

File: theano/tensor/nnet/opt.py (hunk 5 above)
Comment (lamblin) [inline theano/tensor/nnet/opt.py]: (done)

**Comment by lamblin** [comment]
The rebase and PR look good, thanks.
I'm ready to merge when Travis passes.

**Pull request merged**